### PR TITLE
feat: tiktoken integration

### PR DESCRIPTION
**Pull request by nguyenhoangthuan99**
Issue: https://github.com/janhq/cortex.tensorrt-llm/issues/49

### CODE REVIEW

File: cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc [hunk 1]
```diff
@@ -47,12 +48,28 @@ namespace {
...
+
   // TODO(sang) This is fragile, just a temporary solution. Maybe can use a config file or model architect, etc... 
   bool IsOpenhermes(const std::string& s) {
     if (s.find("mistral") != std::string::npos || s.find("Mistral") != std::string::npos) {
       return false;
     } 
     return true;
+  }
+    ModelType GetModelType(const std::string& s){
+    if (s.find("Meta") != std::string::npos || s.find("meta") != std::string::npos) {
```
Comment (vansangpfiev) [inline cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc]: Should use llama3 or Llama3 for this.
We use that name in HF: https://huggingface.co/cortexso/llama3
I know this is ugly, we need to put all these things to the config file soon.

File: cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.h [hunk 2]
```diff
@@ -119,7 +190,8 @@ class TensorrtllmEngine : public EngineI {
   void GetModelStatus(
       std::shared_ptr<Json::Value> json_body,
       std::function<void(Json::Value&&, Json::Value&&)>&& callback) final;
-
+  virtual std::vector<int> encode_header_llama3(const std::string& role);
```
Comment (vansangpfiev) [inline cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.h]: should rename the function to `EncodeHeaderLlama3`, similar for `EncodeMessageLlama3`

File: cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc [hunk 3]
```diff
@@ -220,6 +243,22 @@ bool TensorrtllmEngine::CheckModelLoaded(std::function<void(Json::Value&&, Json:
   return true;
 }
 
+std::vector<int> TensorrtllmEngine::encode_header_llama3(const std::string& role){
+  std::vector<int> tokens = {};
+  tokens.push_back(128006); // <|start_header_id|>
```
Comment (vansangpfiev) [inline cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc]: Put all magic numbers into an enum class?

File: cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc [hunk 4]
```diff
@@ -236,58 +275,73 @@ void TensorrtllmEngine::HandleChatCompletion(std::shared_ptr<Json::Value> json_b
 
   // tokens for Mistral v0.3
   // TODO(sang): too much hard code here, need to refactor it soon
-  std::vector<int32_t> tokens = {static_cast<int32_t>(MistralTemplate::kBos)};
-
+  std::vector<int32_t> tokens;
+  if (model_type_ == ModelType::llama3){
+    tokens.push_back(128000); // <|begin_of_text|>
+  }
```
Comment (vansangpfiev) [inline cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc:294]: else branch for mistral?

File: cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.h [hunk 5]
```diff
@@ -40,32 +83,59 @@ class Tokenizer {
...
+
+  std::vector<int> Encode(const std::string& input) override {
+    std::vector<int> ids = encoder->encode(input);
+    return ids;
+  }
+};
+  enum class ModelType {
+    openhermes,
+    llama3, 
+    mistral,
```
Comment (vansangpfiev) [inline cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.h]: I think should rename these to kOpenHermes, kLlama3, kMistral

File: cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc (hunk 4 above)
Comment (nguyenhoangthuan99) [reply cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc:294]: This logic is only for llama3 chat format. I followed this function https://github.com/meta-llama/llama3/blob/main/llama/tokenizer.py#L222 , they always put "begin_of_text" to the head of tokens, then loop over the list of messages and decode them. 
The mistral doesn't have that logic, it encodes every messages individually.

Comment (vansangpfiev) [reply cpp/tensorrt_llm/cortex.tensorrt-llm/src/tensorrt-llm_engine.cc:294]:
```
std::vector<int32_t> tokens = {static_cast<int32_t>(MistralTemplate::kBos)};
```
The previous version initialized tokens vector with kBos for Mistral, we still need to keep that logic

**Pull request merged**